Always request version.json from network (no cache)

kidroca opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
As explained in the [issue comment](https://github.com/Expensify/App/issues/7856#issuecomment-1082409268) the only usage for `HttpUtils.download` was in the version check here
Instead of modifying existing code to make `HttpUtils.download` support a cache/no-cache options, we use `fetch` directly 
Since now `HttpUtils.download` has no more usages it can be deleted
A [problem with the webpack dev](https://expensify.slack.com/archives/C01GTK53T8Q/p1648595359709439) config was fixed so that the fix can be tested locally

### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/7856

### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

To test locally modify `src/setup/platformSetup/index.website.js` so that this 
https://github.com/Expensify/App/blob/1c4add7b7b691724edd2040110d2adc2b92e7b60/src/setup/platformSetup/index.website.js#L63-L66

becomes:
```js
    // When app loads, get current version (production only)
    checkForUpdates(webUpdater());
```

1. `dist/version.json` is created automatically by the webpack dev server - you should have this file after running `npm run web`
2. Background the app or switch to another tap
3. Switch back to App

- [ ] Verify that no errors appear in the JS console
- [ ] Verify a network request is being made to fetch `./version.json`
- [ ] Verify nothing happens since the fetched version is the same as the `currentVersion`

1. Modify `dist/version.json` to be a greater version than `currentVersion`
2. Background the app or switch to another tap
3. Switch back to App

- [ ] Verify that no errors appear in the JS console
- [ ] Verify a network request is being made to fetch `./version.json`
    - since we expect App to refresh, because a new version is available we should, toggle to preserve Network log between refreshes
    - alternatively you can put a breakpoint inside the code that receive the new version, and code should be paused there after you switch back to the tab. This would also help catch the "Update Available" prompt, since now App would be in the foreground
-  Depending on how fast you switch back and forth between tabs, App can either refresh on the background, or prompt you to confirm refreshing. See the videos under Web that showcase both cases
   - [ ] Switching between tabs fast should show you an "Update available prompt like this" 
    ![image](https://user-images.githubusercontent.com/12156624/160724236-1539a94e-3cb1-4f65-9788-68c84223076a.png)
   - [ ] Backgrounding the app and waiting for a few seconds should initiate a refresh in the background - you should be able to see a spinner briefly appearing on the collapsed tab

### PR Review Checklist
<!--
This is a checklist for PR authors & reviewers. Please make sure to complete all tasks and check them off once you do, or else Expensify has the right not to merge your PR!
-->
#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.


<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
4. Upload an image via copy paste
5. Verify a modal appears displaying a preview of that image
--->

I think this ticket can only be tested after deploying to staging, because we need to be on older version and a new version be released in the meantime

We're primarily verifying there are no regressions and the update check works as expected
All the scenarios below are existing flows and not introduced in the current PR
We're verifying they work as expected

On web/mobile-web

- [ ] Verify new version is detected during background the app
    1. Have App opened and foregrounded before an update is released
    3. Wait for update to be released and then background App - open a new tab
    4. Observer App's tab is refreshing (little spinning circle appears on the tab in Chrome)
    5. Switching back to App's tab you should be on the update version

- [ ] Verify new version is detected during foregrounding the app
    - _This happens only if you're background very quickly (see attached video)_
    1. Have App opened and in the **background** before an update is released
    2. Wait for update to be released and then foreground App
    3. You should see a prompt like this:
       ![image](https://user-images.githubusercontent.com/12156624/160724236-1539a94e-3cb1-4f65-9788-68c84223076a.png)
    4. Pressing "OK" should refresh the page
    6. You should now be on the updated version

- [ ] Verify new version is detected during foregrounding the app - alternative
   - _This happens only if you're background very quickly (see attached video)_
   - Do the 1,2,3 from above, but instead of pressing "OK" on the prompt press "Cancel"
   - The page should not refresh
   - If you background App now it should refresh in the background and the next time you bring it up it would have the updated version

- [ ] Verify problems due to cache are fixed (the bug in the linked ticket)
    1. Shortly (1-2 minutes, up to 30) after an update have been published (on staging or prod) open the web page
    2. Verify you're on the latest version
    3. Verify that opening a new tab or backgrounding the app, does not result in App refresh - app doesn't incorrectly think there's an update 
    4. Verify that opening a new tab or backgrounding the app, doesn't result in showing the "Update available prompt"

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->
- _Note: I've setup my local environment to simulate there's always a new version_

##### New version refreshes in the background
https://user-images.githubusercontent.com/12156624/160727009-84a67faa-5505-41dd-b4fd-777652778998.mov

##### New version prompts to refresh page on the foreground
https://user-images.githubusercontent.com/12156624/160727058-e47da2ac-4a75-4ab8-a2a6-0ecb5d28840d.mov

#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->

##### New version available and refreshes App
https://user-images.githubusercontent.com/12156624/160727223-9dfe6454-f044-4f6b-b713-78014179a8d4.mov

##### No new version available - no page refresh
https://user-images.githubusercontent.com/12156624/160727253-0cd81de8-3cc4-46a3-9c33-b0219dea66e7.mov

##### Android Chrome
https://user-images.githubusercontent.com/12156624/160727885-b72f35c4-f784-439f-9c81-675953d436f6.mp4

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->
![image](https://user-images.githubusercontent.com/12156624/160727659-09f377dd-9a4c-430f-9677-3daa4a10c1fb.png)

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->
![image](https://user-images.githubusercontent.com/12156624/160727325-4a3e770a-34a1-444f-9ea5-aadacb7f638c.png)

#### Android
<!-- Insert screenshots of your changes on the Android platform-->
![image](https://user-images.githubusercontent.com/12156624/160727941-6897998f-14e1-4ac2-8fdf-03ebb2a2e460.png)

kidroca committed:
Download `version.json` ignoring cache using `fetch`

kidroca committed:
Remove `HttpUtils.download` - no more usages

kidroca committed:
Fix webpack.dev contentBase path

kidroca committed:
Update fetched url to `/version.json` so it works locally and on prod

kidroca commented:
I haven't checked some PR check items, because I'm not certain what I should do. 
Waiting to be clarified in this slack thread: https://expensify.slack.com/archives/C01GTK53T8Q/p1648583058174529

luacmartins reviewed (commented):
Hmm I'm following the test steps, I changed the version in `dist/version.json` from `{"version":"1.1.46-3"}` to `{"version":"1.1.47-3"}` but I don't get the prompt to update the app nor do I see a new request for version.json. Am I missing something?

https://user-images.githubusercontent.com/22219519/160890684-670b5574-d4ba-4a61-89a3-1ce9433c944b.mov

kidroca commented:
@luacmartins 
> Hmm I'm following the test steps, I changed the version in `dist/version.json` from `{"version":"1.1.46-3"}` to `{"version":"1.1.47-3"}` but I don't get the prompt to update the app nor do I see a new request for version.json. Am I missing something?

About the prompt it doesn't always appear - I've updated the test steps for QA, and I thought I did for DEV, but apparently not

Regarding the request not appearing in the network tab - it's because App refreshes in the background and you have to enable the [x] preserve log toggle here:
![image](https://user-images.githubusercontent.com/12156624/160904962-31e2c84f-cca9-4b1e-9315-8be1799adfe6.png)

1. App sees there's a newer version
2. App sees you've backgrounded it and it refreshes (which "applies" the update) 
3. Refreshing clears the Network tab

I'll update the Test steps, to match QA's

kidroca commented:
@luacmartins 
I've updated the description and the Test steps

luacmartins commented on src/setup/platformSetup/index.website.js:
```diff
@@ -15,7 +14,8 @@ import Visibility from '../../libs/Visibility';
 function webUpdate() {
-    HttpUtils.download('version.json')
+    fetch('/version.json', {cache: 'no-cache'})
+        .then(response => response.json())
```
Should we check that the response is ok before trying to return json? Something similar to what we do [here](https://github.com/Expensify/App/blob/d89a148655e9d2b99205bae8f6155c94a41e5917/src/libs/HttpUtils.js#L33-L36). Otherwise we might get errors like the one below:

![Screen Shot 2022-03-30 at 10 45 05 AM](https://user-images.githubusercontent.com/22219519/160911242-fcc6595d-d325-448a-b7a1-d1d1710b7506.png)

kidroca replied:
Sorry but it doesn't make sense to me 

We either check for `ok` and do nothing when it's not, which results only in skipping a valid error log in the console
Or we check for `ok` and throw an error ourselves

If we're not going to do something (in webUpdate) when the response is not `ok` it doesn't make sense to add handling
We'll at least see an error message when a request fails for some reason

If we want to do something when an error happens, we should add a `catch` block - it would capture any error, but I just don't see what we could do in such case - capture a server log "update check failed?"

kidroca replied:
We didn't have any error handling for the error that would be thrown here https://github.com/Expensify/App/blob/d89a148655e9d2b99205bae8f6155c94a41e5917/src/libs/HttpUtils.js#L33-L36

So we don't have any error handling now as well

luacmartins replied:
Makes sense! Thanks for clarifying!

luacmartins reviewed (commented):
Thanks for the updated testing steps! Left a small comment.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/2097678131) to staging by @luacmartins in version: 1.1.51-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

mvtglobally commented:
@kidroca @luacmartins @roryabraham 
We had hard time validating this issue as we dont see the exact time the deploy is happening and team tried our best to stop the build changes across devices. 
So far we could only verify Desktop and it looks pass. Web, mWeb and Native apps did not have proper transition and it may be due to the build distribution. 
Would you be able to cross check this internally?

roryabraham commented:
First off – this only affects web AFAIK not other platforms

roryabraham commented:
I went ahead and ran this snippet on staging web:

<img width="364" alt="image" src="https://user-images.githubusercontent.com/47436092/162036939-c6ed4077-61d8-4179-bd7f-b5e9a6c220b3.png">

And honestly, that's good enough for me. I'm going to check this off on the checklist.

roryabraham commented:
We'll go ahead and assume that the [documentation for fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#supplying_request_options) is accurate, and that it's so fundamental to JavaScript that we don't need to do anything more to validate that it's working.

kidroca commented:
Yes, this should only affect web
I think the only way to test it in prod is to wait for an update, if there's some schedule or someone you can contact to synchronize the test it would help 
Otherwise I would just keep a incognito browser open, then wait to be notified that a `staging/prod` build was deployed (https://github.com/Expensify/App/pull/8350#issuecomment-1089144261) and go to that incognito window

If there's a slack channel that posts a message when a build is starting it can be used to open web while the old version is still actual and then do the test steps after the deploy finishes

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/2104164410) to production by @roryabraham in version: 1.1.51-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/2104164410) to production by @roryabraham in version: 1.1.51-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|cancelled 🔪 
🕸 web 🕸|success ✅
